File: Arrays/leaders_array.cpp
#include<iostream>
#include<vector>
#include<algorithm>
#include<climits>
using namespace std;

vector<int>superiorElement_brute(vector<int>&a){
    vector<int>ans;
    for(int i=0;i<a.size();i++){
        bool leader=true;
        for(int j=i+1;j<a.size();j++){
            if(a[j]>a[i]){
                leader=false;
                break;
            }
        }
        if(leader) ans.push_back(a[i]);
    }
    return ans;
}
vector<int> superiorElement(vector<int>&a){
    vector<int>ans;
    int maxi=INT_MIN;
    for(int i=a.size()-1;i>=0;i--){
        if(a[i]>maxi){
            ans.push_back(a[i]);
        }
        maxi=max(maxi,a[i]);
    }
    reverse(ans.begin(), ans.end());
    return ans;
}


int main() {
    vector<int> a = {1, 2, 3, 2, 1, 5, 3};
    vector<int> res = superiorElement(a);

    cout << "Superior elements are: ";
    for (int x : res) {
        cout << x << " ";
    }
    cout << endl;

    return 0;
} 

File: java/ArrayQ/LeadersArray.java
import java.util.*;

public class LeadersArray{

    // ---------------- Brute Force ----------------
    // Time Complexity: O(N^2)
    // Space Complexity: O(1) (excluding output list)
    static ArrayList<Integer> superiorElementBrute(int[] arr) {
        ArrayList<Integer> ans = new ArrayList<>();
        int n = arr.length;

        for (int i = 0; i < n; i++) {
            boolean leader = true;
            for (int j = i + 1; j < n; j++) {
                if (arr[j] > arr[i]) {
                    leader = false;
                    break;
                }
            }
            if (leader) ans.add(arr[i]);
        }
        return ans;
    }

    // ---------------- Optimized ----------------
    // Time Complexity: O(N)
    // Space Complexity: O(1) (excluding output list)
    static ArrayList<Integer> superiorElement(int[] arr) {
        ArrayList<Integer> ans = new ArrayList<>();
        int n = arr.length;

        int maxi = Integer.MIN_VALUE;

        for (int i = n - 1; i >= 0; i--) {
            if (arr[i] > maxi) {
                ans.add(arr[i]);
            }
            maxi = Math.max(maxi, arr[i]);
        }

        // Reverse to maintain original order
        Collections.reverse(ans);
        return ans;
    }

    public static void main(String[] args) {
        int[] arr = {1, 2, 3, 2, 1, 5, 3};

        ArrayList<Integer> res = superiorElement(arr);

        System.out.print("Superior elements are: ");
        for (int x : res) {
            System.out.print(x + " ");
        }
    }
}
